### PR TITLE
[Data Team] ML model pouring enhancements

### PR DESCRIPTION
**Pull request by RobertXu**
### Description

This PR adds a ML model for detecting cup fullness, buttons for labeling a pour and an automatic retraining loop that allows users to capture images, train and promote ML models all without interacting with the UI. 

#### ML Pour Detection

- New use_ml config field on Config to toggle between the existing image-delta logic and an ML model (DataFullnessService) for stop-detection during a pour.
- The model was trained on a dataset of water, wine and colored liquids. 

#### Model Retraining Loop

- During each pour, images are saved to a timestamped subdirectory under training/
- After the pour, the user can press a stream deck button that triggers a do-command for `good-pour`, `under-pour` or `over-pour`. This labels the images as `full/not-full` and uploads the images to the cropped-cup dataset. 
- If the user presses the labeling buttons > 15 mins after the last pour, we do not label and upload the images. This ensures that 

#### Automatic Training Script (training-script/)

- This script is run once a week by a cron job. It trains a new version of the model and compares it against a source of truth dataset. It updates the model on the machine if the model passes an accuracy threshold.

### CODE REVIEW

File: cmd/tools/cmd-tools.go [hunk 1]
```diff
@@ -181,6 +181,8 @@ func realMain() error {
 		}
 		defer file.Close()
 		return png.Encode(file, img)
+	case "get-glass-quickly":
```
Comment (RobertXu) [inline cmd/tools/cmd-tools.go]: remove

File: pour/vinocart.go [hunk 2]
```diff
@@ -44,8 +47,16 @@ var vinowebStaticFS embed.FS
...
 const gripperToCupCenterHack = -35
 
+const goodPour = "good-pour"
+const underPour = "under-pour"
+const overPour = "over-pour"
+const trainingDataDirName = "training"
+
 var VinoCartModel = NamespaceFamily.WithModel("vinocart")
 var noObjects = fmt.Errorf("no objects")
+var bottlePouringDatasetID = "696a83481a9adaa357b3c8b8"
```
Comment (RobertXu) [inline pour/vinocart.go]: remove

File: pour/vinocart.go [hunk 3]
```diff
@@ -201,6 +256,30 @@ func (vc *VinoCart) DoCommand(ctx context.Context, cmd map[string]interface{}) (
 		return nil, vc.PutBack(ctx)
 	}
 
+	if cmd["get-glass-quickly"] == true {
```
Comment (RobertXu) [inline pour/vinocart.go]: remove

Comment (RobertXu) [reply pour/vinocart.go]: go over the other do commands

File: pour/vinocart.go [hunk 4]
```diff
@@ -410,7 +559,7 @@ func (vc *VinoCart) checkPickQuality(ctx context.Context) error {
 
 	if vc.dataClient != nil {
 		vc.logger.Infof("uploading image to dataset for cup pick quality")
-		err := saveImageToDataset(ctx, vc.c.Cam.Name(), prepped, vc.dataClient, "683f8952383a821481d9b5c9")
+		err := saveImageToDataset(ctx, vc.c.Cam.Name(), prepped, vc.dataClient, "683f8952383a821481d9b5c9", nil)
```
Comment (RobertXu) [inline pour/vinocart.go]: Move id to a constant

File: pour/vinocart.go [hunk 5]
```diff
@@ -800,6 +1066,14 @@ func (vc *VinoCart) PourPrep(ctx context.Context) error {
 		return err
 	}
 
+	// wait for user to select model mode before continuing
```
Comment (RobertXu) [inline pour/vinocart.go]: go over how this works in practice

File: pour/vinocart.go [hunk 6]
```diff
@@ -860,6 +1134,20 @@ func (vc *VinoCart) moveToCurrentXYAtCupHeight(ctx context.Context) error {
 	return err
 }
 
+func (vc *VinoCart) getPourDetails(ctx context.Context, img image.Image) (string, float64, error) {
```
Comment (RobertXu) [inline pour/vinocart.go]: Can change this back to returning a true/false (`shouldStopPour`) vs the additional debugging info.

File: pour/vinocart.go [hunk 7]
```diff
@@ -940,18 +1229,16 @@ func (vc *VinoCart) Pour(ctx context.Context) error {
...
 	time.Sleep(500 * time.Millisecond)
 
-	start := time.Now()
-	loopNumber := 0
-
+	// oldModel vars
 	var pd *pourDetector
+	markedDifferent := false
 
 	totalTime := 15 * time.Second
```
Comment (RobertXu) [inline pour/vinocart.go]: Look into having a shorter maxTime if we are using the ML model to prevent overpours.

File: pour/vinocart.go [hunk 8]
```diff
@@ -960,7 +1247,7 @@ func (vc *VinoCart) Pour(ctx context.Context) error {
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
-			err := saveImageToDatasetFromCamera(context.Background(), vc.c.GlassPourCam, vc.dataClient, "683d1210c83b3f3823ec70ff")
+			err := saveImageToDatasetFromCamera(context.Background(), vc.c.GlassPourCam, vc.dataClient, "696a83481a9adaa357b3c8b8")
```
Comment (RobertXu) [inline pour/vinocart.go]: move dataset id to a variable

File: pour/vinocart.go [hunk 9]
```diff
@@ -994,44 +1280,124 @@ func (vc *VinoCart) Pour(ctx context.Context) error {
...
+	// create directory where we save images, named after start time
+	pourTime := ""
+	pourTime = time.Now().Format("20060102_150405.000")
+	subDir := filepath.Join(trainingDataDirName, pourTime)
+	if err := os.Mkdir(subDir, 0o755); err != nil {
+		return err
+	}
+	vc.imgDirName = subDir
+
+	timeoutReached := true
```
Comment (RobertXu) [inline pour/vinocart.go]: can remove, was just for debugging.

File: pour/vinocart.go [hunk 10]
```diff
@@ -994,44 +1280,124 @@ func (vc *VinoCart) Pour(ctx context.Context) error {
...
+			isFull, err := vc.handleMLPour(ctx, img, loopNumber)
+			if err != nil {
+				return err
+			}
+			if isFull {
+				timeoutReached = false
+				break PourLoop
+			}
+		case !useMLModel:
+			vc.logger.Info("*** using old pour model ***")
```
Comment (RobertXu) [inline pour/vinocart.go]: change wording to talk about "image delta logic" instead of "old pour model".

File: pour/vinocart.go (hunk 2 above)
Comment (RobertXu) [reply pour/vinocart.go]: don't remove, use as variable

Comment (RobertXu) [reply pour/vinocart.go]: jk, remove

File: pour/vinocart.go [hunk 11]
```diff
@@ -201,6 +256,30 @@ func (vc *VinoCart) DoCommand(ctx context.Context, cmd map[string]interface{}) (
 		return nil, vc.PutBack(ctx)
 	}
 
+	if cmd["get-glass-quickly"] == true {
+		return nil, vc.GetGlassQuickly(ctx)
+	}
+
+	if cmd["tilt-bottle-forward"] == true {
```
Comment (RobertXu) [inline pour/vinocart.go]: remove

File: pour/vinocart.go [hunk 12]
```diff
@@ -201,6 +256,30 @@ func (vc *VinoCart) DoCommand(ctx context.Context, cmd map[string]interface{}) (
...
+		return nil, vc.TiltBottleForward(ctx)
+	}
+
+	if cmd["tilt-bottle-backward"] == true {
+		return nil, vc.TiltBottleBackward(ctx)
+	}
+
+	if cmd["return-bottle-to-pre-pour-position"] == true {
```
Comment (RobertXu) [inline pour/vinocart.go]: remove

File: pour/vinocart.go [hunk 13]
```diff
@@ -201,6 +256,30 @@ func (vc *VinoCart) DoCommand(ctx context.Context, cmd map[string]interface{}) (
...
+		return nil, vc.ReturnBottleToPrePourPosition(ctx)
+	}
+
+	if cmd["capture-not-full"] == true {
```
Comment (RobertXu) [inline pour/vinocart.go]: remove

File: pour/vinocart.go [hunk 14]
```diff
@@ -223,6 +302,52 @@ func (vc *VinoCart) DoCommand(ctx context.Context, cmd map[string]interface{}) (
...
+
+	if cmd["under-pour"] == true {
+		return nil, vc.labelPour(ctx, underPour)
+	}
+
+	if cmd["good-pour"] == true {
+		return nil, vc.labelPour(ctx, goodPour)
+	}
+
+	if cmd["start-capture"] == true {
```
Comment (RobertXu) [inline pour/vinocart.go]: remove both

File: pour/vinocart.go [hunk 15]
```diff
@@ -994,44 +1280,124 @@ func (vc *VinoCart) Pour(ctx context.Context) error {
...
+	vc.logger.Infow("Got start timestamp for pour loop", "startTime", start.String())
+	loopNumber := 0
+
+PourLoop:
 	for time.Since(start) < totalTime {
 		loopStart := time.Now()
 
-		img, fn, err := vc.DebugGetGlassPourCamImage(ctx /* loopNumber */, box, -1)
+		// get image
+		img, fn, err := vc.DebugGetGlassPourCamImage(ctx /* loopNumber */, box, loopNumber)
```
Comment (RobertXu) [inline pour/vinocart.go]: Change naming of this function, b/c ML model uses this in non-debug way.

File: pour/config.go [hunk 16]
```diff
@@ -88,6 +97,10 @@ func (cfg *Config) Validate(path string) ([]string, []string, error) {
 		deps = append(deps, cfg.PourGlassFindService)
 	}
 
+	if cfg.DataFullnessService != "" {
```
Comment (RobertXu) [inline pour/config.go]: This wraps around the new ML model for determing cup fullness. It can be toggled on/off using the `UseMLModel` config boolean.

File: pour/vinocart.go [hunk 17]
```diff
@@ -3,13 +3,15 @@ package pour
 import (
 	"context"
 	"embed"
-	"encoding/json"
+	"errors"
```
Comment (RobertXu) [inline pour/vinocart.go:6]: There's a lot of small tweaks happening in this file. Happy to explain anything in more detail.

File: pour/config.go [hunk 18]
```diff
@@ -73,8 +73,15 @@ type Config struct {
 
 	PickQualityService   string `json:"pick_quality_service"`
 	PourGlassFindService string `json:"pour_glass_find_service"`
+	DataFullnessService  string `json:"data_fullness_service"`
+
+	// pass in the org api key for data client
+	APIKey   string `json:"api_key"`
```
Comment (erh) [inline pour/config.go]: you can get these from the environment

File: pour/config.go [hunk 19]
```diff
@@ -73,8 +73,15 @@ type Config struct {
...
+	APIKeyID string `json:"api_key_id"`
 
 	Loop bool `json:"loop"`
+
+	UseMLModel bool `json:"use_ml"`
```
Comment (erh) [inline pour/config.go]: not specific enough

File: pour/config.go [hunk 20]
```diff
@@ -176,6 +187,7 @@ type Pour1Components struct {
 
 	PickQualityService   vision.Service
 	PourGlassFindService vision.Service
+	DataFullnessService  vision.Service
```
Comment (erh) [inline pour/config.go]: isn't it glassfullness?

File: pour/vinocart.go [hunk 21]
```diff
@@ -68,7 +80,35 @@ func newVinoCart(ctx context.Context, deps resource.Dependencies, conf resource.
 		return nil, err
 	}
 
-	g, err := NewVinoCart(ctx, config, c, robotClient, nil, logger)
+	var dataClient *app.DataClient
+	var appClient *app.ViamClient
+	if config.APIKey != "" && config.APIKeyID != "" {
```
Comment (erh) [inline pour/vinocart.go]: get rid of this path

File: pour/vinocart.go [hunk 22]
```diff
@@ -130,6 +171,9 @@ func NewVinoCart(ctx context.Context, conf *Config, c *Pour1Components, client r
 		vc.server.ListenAndServe()
 	}()
 
+	vc.pourStep = 0
+
+	vc.useMLModel = conf.UseMLModel
```
Comment (erh) [inline pour/vinocart.go]: why?

File: pour/vinocart.go [hunk 23]
```diff
@@ -157,6 +202,14 @@ type VinoCart struct {
 	status     string
 
 	server *http.Server
+
+	pourStep int
```
Comment (erh) [inline pour/vinocart.go]: this seems unused

File: pour/vinocart.go [hunk 24]
```diff
@@ -1133,44 +1333,100 @@ func (vc *VinoCart) Pour(ctx context.Context) error {
...
 	}
 
 	// cleanup done in defer above
+	vc.logger.Infow(" **** pour loop done **** ", "elapsed time", time.Since(start).String())
 	return nil
 }
 
-func saveImage(img image.Image, loopNumber int) (string, error) {
-	fn := fmt.Sprintf("img-%d.png", loopNumber)
+func (vc *VinoCart) handleMLPour(ctx context.Context, img image.Image, loopNumber int) (bool, error) {
```
Comment (erh) [inline pour/vinocart.go]: this is really just "isFullUsingMLModel" right?
i don't know what handle means?

File: pour/vinocart.go [hunk 25]
```diff
@@ -1133,44 +1333,100 @@ func (vc *VinoCart) Pour(ctx context.Context) error {
...
+
+	if vc.useMLModel {
+		vc.logger.Info("*** using ml pour model ***")
+	} else {
+		vc.logger.Info("*** using image delta logic ***")
+	}
+
+	start := time.Now()
+	vc.logger.Infow("Got start timestamp for pour loop", "startTime", start.String())
+	loopNumber := 0
```
Comment (erh) [inline pour/vinocart.go]: all of this doesn't need to change

File: training-script/cmd/main.go [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+	"os"
+
+	// Blank-import the function package so the init() runs
+	"github.com/GoogleCloudPlatform/functions-framework-go/funcframework"
+	_ "go.viam.com/trainingscript"
+)
+
+// This is for local testing. to deploy the cloud function, run: gcloud run deploy  retraining-model-script --source . --function TrainModelHTTP --base-image go125 --region us-central1 --project staging-cloud-web-app
+
+func main() {
```
Comment (erh) [inline training-script/cmd/main.go]: i'm very confused what this is here for

File: training-script/go.mod [hunk 27]
```diff
@@ -0,0 +1,140 @@
+module go.viam.com/trainingscript
```
Comment (erh) [inline training-script/go.mod]: need to make directory more specific, what training script

File: training-script/test_model.go [hunk 28]
```diff
@@ -0,0 +1,173 @@
+package trainingscript
```
Comment (erh) [inline training-script/test_model.go]: why can't we use the viam built in model building for this?

**Review by erh — CHANGES_REQUESTED** [review]

File: pour/config.go (hunk 20 above)
Comment (RobertXu) [reply pour/config.go]: Yep, the boolean and service have been renamed to be more descriptive.

File: training-script/cmd/main.go (hunk 26 above)
Comment (RobertXu) [reply training-script/cmd/main.go]: I should have made the purpose of the training script clearer. It's a fyi about how the retraining loop works. The actual code doesn't live in the module though since it runs through a cloud function. I've removed the training script to reflect that.

File: training-script/test_model.go (hunk 28 above)
Comment (RobertXu) [reply training-script/test_model.go]: This would be a useful enhancement to our model training setup. The purpose of this script is to evaluate/score a model against a predefined dataset with source of truth labels. If the model passes an accuracy threshold we then add it to the wine cart.

**Comment by dmhilly** [comment]
Hey @erh, FYI that I'm taking over this PR. LMK if you have remaining questions or concerns
